src/compiler/scala/tools/nsc/symtab/classfile/ReusableDataReader.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.nsc.symtab.classfile
+
+import java.io.{ByteArrayInputStream, DataInputStream, InputStream}
+import java.nio.channels.Channels
+import java.nio.{BufferUnderflowException, ByteBuffer}
+
+final class ReusableDataReader() extends DataReader {
+  private[this] var data = new Array[Byte](32768)
+  private[this] var bb: ByteBuffer = ByteBuffer.wrap(data)
+  private[this] var size = 0
+  private[this] val reader: DataInputStream = {
+    val stream = new InputStream {
+      override def read(): Int = try {
```

Comment:
It would be better to override also the default implementation of `public int read(byte b[], int off, int len)` because it is used in `readFully` immediately and can save a lot of CPU cycles.

Reply:
Good idea!